File: src/engine/transforms.ts
import type { FieldValue } from '../widgets/contracts/fields'
import type { WireTransform } from '../types/circuit'

// ---------------------------------------------------------------------------
// Wire transforms — pure, total functions over field values.
//
// A transform never throws and never returns NaN/Infinity: a wire must be
// unable to poison the widget it feeds. Coercion mirrors the tolerant
// readers in src/widgets/fields.ts so the two layers agree on meaning.
// ---------------------------------------------------------------------------

function toNumber(value: FieldValue): number {
  if (Array.isArray(value)) return value.at(-1)?.v ?? 0
  if (typeof value === 'number') return Number.isFinite(value) ? value : 0
  if (typeof value === 'boolean') return value ? 1 : 0
  const parsed = parseFloat(value)
  return Number.isFinite(parsed) ? parsed : 0
}

function toText(value: FieldValue): string {
  if (Array.isArray(value)) return value.map((point) => point.v).join(', ')
  return typeof value === 'string' ? value : String(value)
}

function toBool(value: FieldValue): boolean {
  if (Array.isArray(value)) return value.length > 0
  if (typeof value === 'boolean') return value
  if (typeof value === 'number') return value >= 1
  return value === 'true' || value === '1' || value === 'yes' || value === 'on'
}

function finite(value: number): number {
  return Number.isFinite(value) ? value : 0
}

/** Apply a wire transform. `undefined`/identity passes the value through. */
export function applyTransform(value: FieldValue, transform?: WireTransform): FieldValue {
  if (!transform) return value
  switch (transform.op) {
    case 'identity':
      return value
    case 'scale':
      return finite(toNumber(value) * transform.factor)
    case 'offset':
      return finite(toNumber(value) + transform.amount)
    case 'clamp': {
      const lo = Math.min(transform.min, transform.max)
      const hi = Math.max(transform.min, transform.max)
      return Math.min(hi, Math.max(lo, toNumber(value)))
    }
    case 'map_range': {
      const span = transform.inMax - transform.inMin
      if (span === 0) return finite(transform.outMin)
      const t = (toNumber(value) - transform.inMin) / span
      return finite(transform.outMin + t * (transform.outMax - transform.outMin))
    }
    case 'round':
      return Math.round(toNumber(value))
    case 'invert':
      return typeof value === 'boolean' ? !value : finite(-toNumber(value))
    case 'threshold':
      return toNumber(value) >= transform.value
    case 'format':
      return transform.template.replaceAll('{value}', toText(value))
  }
}

/**
 * Stable serialization for change detection. Field values are primitives or
 * small series arrays, so this stays cheap; the engine compares these strings
 * to decide whether a wire has anything new to say.
 */
export function serializeFieldValue(value: FieldValue): string {
  if (Array.isArray(value)) {
    let out = 's:'
    for (const point of value) out += `${point.t},${point.v};`
    return out
  }
  if (typeof value === 'number') return `n:${value}`
  if (typeof value === 'boolean') return `b:${value}`
  return `t:${value}`
}

/** The boolean reading of a field value — trigger edges are detected on this. */
export function fieldValueAsBool(value: FieldValue): boolean {
  return toBool(value)
}

File: apple/Sources/GrovepadCore/Circuit/Transforms.swift
import Foundation

// ---------------------------------------------------------------------------
// Wire transforms — pure, total functions over field values
// (`engine/transforms.ts`). A transform never throws and never returns
// NaN/Infinity: a wire must be unable to poison the widget it feeds.
// ---------------------------------------------------------------------------

/// `Number.isFinite(value) ? value : 0`.
@inlinable func finite(_ value: Double) -> Double {
    value.isFinite ? value : 0
}

/// Apply a wire transform. `nil` and identity pass the value through.
public func applyTransform(_ value: FieldValue, _ transform: WireTransform?) -> FieldValue {
    guard let transform else { return value }
    switch transform {
    case .identity:
        return value
    case .scale(let factor):
        return .number(finite(num(value) * factor))
    case .offset(let amount):
        return .number(finite(num(value) + amount))
    case .clamp(let a, let b):
        let lo = Swift.min(a, b)
        let hi = Swift.max(a, b)
        return .number(Swift.min(hi, Swift.max(lo, num(value))))
    case .mapRange(let inMin, let inMax, let outMin, let outMax):
        let span = inMax - inMin
        if span == 0 { return .number(finite(outMin)) }
        let t = (num(value) - inMin) / span
        return .number(finite(outMin + t * (outMax - outMin)))
    case .round:
        return .number(jsRound(num(value)))
    case .invert:
        if case .bool(let flag) = value { return .bool(!flag) }
        return .number(finite(-num(value)))
    case .threshold(let threshold):
        return .bool(num(value) >= threshold)
    case .format(let template):
        return .text(JavaScript.replaceAll(template, "{value}", with: text(value)))
    }
}

/// Stable serialization for change detection: the engine compares these
/// strings to decide whether a wire has anything new to say.
public func serializeFieldValue(_ value: FieldValue) -> String {
    switch value {
    case .series(let points):
        var out = "s:"
        for point in points {
            out += JavaScript.numberString(point.t) + "," + JavaScript.numberString(point.v) + ";"
        }
        return out
    case .number(let number): return "n:" + JavaScript.numberString(number)
    case .bool(let flag): return "b:" + (flag ? "true" : "false")
    case .text(let string): return "t:" + string
    }
}

/// The boolean reading of a field value — trigger edges are detected on this.
public func fieldValueAsBool(_ value: FieldValue) -> Bool {
    bool(value)
}
